File: Semestre_2/Vision_Computadora/Entregas/Matlab/T3_2_Gauseana_2D.m
clear;
close all;
clc;

[file,path] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'},'File Selector - Imagen A'); 
ruta=strcat(path,file);
A=imread(ruta);
A= im2double(A);

% Crear filtros 2D  (media y gaussiano)
media11 = fspecial('average', [11 11]);      % Filtro media 11x11
media21 = fspecial('average', [21 21]);      % Filtro media 21x21
gauss21 = fspecial('gaussian', [21 21], 5);  % Filtro gaussiano 21x21 (σ = 5)
gauss41 = fspecial('gaussian', [41 41], 10); % Filtro gaussiano 41x41 (σ = 10)

% Aplicar filtros 
R1 = imfilter(A, media11, 'replicate');   % Media 11x11
R2 = imfilter(A, media21, 'replicate');   % Media 21x21
R3 = imfilter(A, gauss21, 'replicate');   % Gaussiana 21x21
R4 = imfilter(A, gauss41, 'replicate');   % Gaussiana 41x41

% Mostrar resultados
figure;
subplot(2,2,1); imshow(R1); title('Media 11x11');
subplot(2,2,2); imshow(R2); title('Media 21x21');
subplot(2,2,3); imshow(R3); title('Gaussiana 21x21');
subplot(2,2,4); imshow(R4); title('Gaussiana 41x41');

% Crear filtros 1D (horizontales y verticales)
media_horiz = fspecial('average', [1 31]);       % Filtro media horizontal
media_vert  = fspecial('average', [31 1]);       % Filtro media vertical
gauss_horiz = fspecial('gaussian', [1 61], 10);  % Gaussiana horizontal σ=10
gauss_vert  = fspecial('gaussian', [61 1], 10);  % Gaussiana vertical σ=10

% Aplicar filtros 
R5 = imfilter(A, media_horiz, 'replicate');   % Media horizontal
R6 = imfilter(A, media_vert, 'replicate');    % Media vertical
R7 = imfilter(A, gauss_horiz, 'replicate');   % Gaussiana horizontal
R8 = imfilter(A, gauss_vert, 'replicate');    % Gaussiana vertical

%  Mostrar resultados 
figure;
subplot(2,2,1); imshow(R5); title('Media 31x1 (horizontal)');
subplot(2,2,2); imshow(R6); title('Media 1x31 (vertical)');
subplot(2,2,3); imshow(R7); title('Gaussiana 61x1 (horizontal)');
subplot(2,2,4); imshow(R8); title('Gaussiana 1x61 (vertical)');

%  Filtros 2D
tic;  % Inicia cronómetro para filtro de media 11x11
T1 = imfilter(A, media11, 'replicate');  % Aplica el filtro con borde replicado
t_media11 = toc;  % Registra el tiempo transcurrido

tic;  % Filtro de media 21x21
T2 = imfilter(A, media21, 'replicate');
t_media21 = toc;

tic;  % Filtro gaussiano 21x21
T3 = imfilter(A, gauss21, 'replicate');
t_gauss21 = toc;

tic;  % Filtro gaussiano 41x41
T4 = imfilter(A, gauss41, 'replicate');
t_gauss41 = toc;

%  Filtros 1D

tic;  % Filtro media horizontal 1x31
T5 = imfilter(A, media_horiz, 'replicate');
t_mediaH = toc;

tic;  % Filtro media vertical 31x1
T6 = imfilter(A, media_vert, 'replicate');
t_mediaV = toc;

tic;  % Filtro gaussiano horizontal 1x61
T7 = imfilter(A, gauss_horiz, 'replicate');
t_gaussH = toc;

tic;  % Filtro gaussiano vertical 61x1
T8 = imfilter(A, gauss_vert, 'replicate');
t_gaussV = toc;


% Mostrar tiempos 
fprintf('\nTiempos de ejecución (en segundos):\n');
fprintf('Media 11x11         : %.6f\n', t_media11);
fprintf('Media 21x21         : %.6f\n', t_media21);
fprintf('Gaussiana 21x21     : %.6f\n', t_gauss21);
fprintf('Gaussiana 41x41     : %.6f\n', t_gauss41);
fprintf('Media 31x1 (H)      : %.6f\n', t_mediaH);
fprintf('Media 1x31 (V)      : %.6f\n', t_mediaV);
fprintf('Gaussiana 61x1 (H)  : %.6f\n', t_gaussH);
fprintf('Gaussiana 1x61 (V)  : %.6f\n', t_gaussV);

File: Semestre_2/Vision_Computadora/Entregas/Python/T3_2_Gauseana_2D.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from tkinter import Tk, filedialog
from scipy.ndimage import convolve
import time

file_path = filedialog.askopenfilename(
    title='File Selector - Imagen A',
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

A = imread(file_path).astype(float) / 255.0
if A.ndim == 3 and A.shape[-1] == 4:
    A = A[..., :3]

def aplicar_filtro(img, kernel):
    if img.ndim == 3:
        R = np.zeros_like(img)
        for c in range(img.shape[2]):
            R[..., c] = convolve(img[..., c], kernel, mode='nearest')
    else:
        R = convolve(img, kernel, mode='nearest')
    return R

def gauss2d(size, sigma):
    n = size
    ax = np.linspace(-(n-1)/2, (n-1)/2, n)
    xx, yy = np.meshgrid(ax, ax)
    k = np.exp(-(xx**2 + yy**2) / (2 * sigma**2))
    k /= k.sum()
    return k

media11 = np.ones((11, 11)) / (11 * 11)
media21 = np.ones((21, 21)) / (21 * 21)
gauss21 = gauss2d(21, 5)
gauss41 = gauss2d(41, 10)

R1 = aplicar_filtro(A, media11)
R2 = aplicar_filtro(A, media21)
R3 = aplicar_filtro(A, gauss21)
R4 = aplicar_filtro(A, gauss41)

plt.figure()
plt.subplot(2,2,1); plt.imshow(R1); plt.title('Media 11x11'); plt.axis('off')
plt.subplot(2,2,2); plt.imshow(R2); plt.title('Media 21x21'); plt.axis('off')
plt.subplot(2,2,3); plt.imshow(R3); plt.title('Gaussiana 21x21'); plt.axis('off')
plt.subplot(2,2,4); plt.imshow(R4); plt.title('Gaussiana 41x41'); plt.axis('off')

media_horiz = np.ones((1, 31)) / 31
media_vert  = np.ones((31, 1)) / 31

ax = np.linspace(-(61-1)/2, (61-1)/2, 61)
g1d = np.exp(-ax**2 / (2 * 10**2))
g1d /= g1d.sum()
gauss_horiz = g1d.reshape(1, -1)
gauss_vert  = g1d.reshape(-1, 1)

R5 = aplicar_filtro(A, media_horiz)
R6 = aplicar_filtro(A, media_vert)
R7 = aplicar_filtro(A, gauss_horiz)
R8 = aplicar_filtro(A, gauss_vert)

plt.figure()
plt.subplot(2,2,1); plt.imshow(R5); plt.title('Media 31x1 (horizontal)'); plt.axis('off')
plt.subplot(2,2,2); plt.imshow(R6); plt.title('Media 1x31 (vertical)'); plt.axis('off')
plt.subplot(2,2,3); plt.imshow(R7); plt.title('Gaussiana 61x1 (horizontal)'); plt.axis('off')
plt.subplot(2,2,4); plt.imshow(R8); plt.title('Gaussiana 1x61 (vertical)'); plt.axis('off')
plt.show()

t0 = time.time()
T1 = aplicar_filtro(A, media11)
t_media11 = time.time() - t0

t0 = time.time()
T2 = aplicar_filtro(A, media21)
t_media21 = time.time() - t0

t0 = time.time()
T3 = aplicar_filtro(A, gauss21)
t_gauss21 = time.time() - t0

t0 = time.time()
T4 = aplicar_filtro(A, gauss41)
t_gauss41 = time.time() - t0

t0 = time.time()
T5 = aplicar_filtro(A, media_horiz)
t_mediaH = time.time() - t0

t0 = time.time()
T6 = aplicar_filtro(A, media_vert)
t_mediaV = time.time() - t0

t0 = time.time()
T7 = aplicar_filtro(A, gauss_horiz)
t_gaussH = time.time() - t0

t0 = time.time()
T8 = aplicar_filtro(A, gauss_vert)
t_gaussV = time.time() - t0

print('\nTiempos de ejecución (en segundos):')
print(f'Media 11x11         : {t_media11:.6f}')
print(f'Media 21x21         : {t_media21:.6f}')
print(f'Gaussiana 21x21     : {t_gauss21:.6f}')
print(f'Gaussiana 41x41     : {t_gauss41:.6f}')
print(f'Media 31x1 (H)      : {t_mediaH:.6f}')
print(f'Media 1x31 (V)      : {t_mediaV:.6f}')
print(f'Gaussiana 61x1 (H)  : {t_gaussH:.6f}')
print(f'Gaussiana 1x61 (V)  : {t_gaussV:.6f}')
